### DISCUSSION TITLE
Question about quic server creating udp socket based on the number of cpu cores on the machine

### DISCUSSION THREAD

**Discussion by MeristemForge** [microsoft/msquic#2567] [Q&A] [answered]
The quic server creates multiple udp sockets based on the number of cpu cores and binds the same address. If multiple client connections arrive, can the multiple sockets created on the server side respond to their own load balancing?

**Comment by nibanks** [comment]
I'm unsure of exactly what you're asking. How the client connection is handled depends on load balancing and RSS, and it depends on if the client reuses the same UDP 4-tuple or not. Could you please elaborate on the scenario you're describing and exactly what info you're looking for?

**Comment by MeristemForge** [comment, marked as answer]
Creating a server in msquic's code will create multiple udp sockets and they all bind to the same local address, so I wonder if these udp sockets can monitor client connections? 




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "microsoft/msquic"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年3月22日(星期二) 晚上6:13
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [microsoft/msquic] Question about quic server creating udp socket based on the number of cpu cores on the machine (Discussion #2567)





 
I'm unsure of exactly what you're asking. How the client connection is handled depends on load balancing and RSS, and it depends on if the client reuses the same UDP 4-tuple or not. Could you please elaborate on the scenario you're describing and exactly what info you're looking for?

**Reply by nibanks** [reply]
The point of the multiple sockets on the server side is to support [RSS](https://docs.microsoft.com/en-us/windows-hardware/drivers/network/introduction-to-receive-side-scaling).

**Reply by MeristemForge** [reply]
Does msquic also create multiple udp sockets if the network controller doesn't support multiple queues? 
Is it true that only one udp socket works in this case? 
If the network controller supports multiple queues, then whether it is the same as SO_REUSEPORT under linux, it can be load balanced

**Reply by nibanks** [reply]
Yes, MsQuic always creates multiple sockets, because it doesn't know the RSS configuration, so it just assumes it's enabled/supported. Sockets/threads that don't end up mapping to an RSS queue just go unused, and on most modern systems the overhead is inconsequential.

**Reply by MeristemForge** [reply]
Hi nibanks，

Thanks a lot for the help.
